File: 1340-the-dining-philosophers/1340-the-dining-philosophers.java
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

class DiningPhilosophers {

    private final Lock[] forks = new ReentrantLock[5];
    private final Lock table = new ReentrantLock();  // global lock to control philosopher start order

    public DiningPhilosophers() {
        for (int i = 0; i < 5; i++) {
            forks[i] = new ReentrantLock();
        }
    }

    public void wantsToEat(int philosopher,
                           Runnable pickLeftFork,
                           Runnable pickRightFork,
                           Runnable eat,
                           Runnable putLeftFork,
                           Runnable putRightFork) throws InterruptedException {
        int left = philosopher;
        int right = (philosopher + 1) % 5;

        table.lock(); // enforce strict order by allowing only one philosopher to attempt fork-pick at a time
        forks[right].lock();
        forks[left].lock();
        try {
            // Pick in platform-required order: right then left
            pickRightFork.run();
            pickLeftFork.run();

            eat.run();

            // Put down in platform-required order: left then right
            putLeftFork.run();
            putRightFork.run();
        } finally {
            forks[left].unlock();
            forks[right].unlock();
            table.unlock();
        }
    }
}

File: 1340-the-dining-philosophers/1340-the-dining-philosophers.py
import threading
from typing import Callable

class DiningPhilosophers:

    def __init__(self):
        # Create 5 locks, one for each fork
        self.forks = [threading.Lock() for _ in range(5)]

    def wantsToEat(self,
                   philosopher: int,
                   pickLeftFork: Callable[[], None],
                   pickRightFork: Callable[[], None],
                   eat: Callable[[], None],
                   putLeftFork: Callable[[], None],
                   putRightFork: Callable[[], None]) -> None:
        
        left = philosopher
        right = (philosopher + 1) % 5

        # To prevent deadlock, always acquire the lower-numbered fork first
        first, second = (left, right) if left < right else (right, left)

        # Get the locks corresponding to the forks
        first_lock = self.forks[first]
        second_lock = self.forks[second]

        # Use nested context managers to lock both forks
        with first_lock:
            with second_lock:
                # Pick up forks in correct order
                if first == left:
                    pickLeftFork()
                    pickRightFork()
                else:
                    pickRightFork()
                    pickLeftFork()

                # Eat
                eat()

                # Put down forks in the same order
                if first == left:
                    putLeftFork()
                    putRightFork()
                else:
                    putRightFork()
                    putLeftFork()

        